mmpose/core/evaluation/top_down_eval.py

```diff
@@ -250,6 +250,83 @@ def _taylor(heatmap, coord):
+
+    Note:
+        batch_size: N
+        num_keypoints: K
+
+    Args:
+        coords (np.ndarray[N, K, 2]): Coords in shape of batchsize*num_kps*2.
+        batch_heatmaps (np.ndarray[N, K, H, W]): batch_heatmaps
+        share_heatmap (bool) : True for bottom_up paradigm where all persons
+                    share the same heatmap while False for top_down paradigm
+                    where each person has its own heatmaps.
+
+    Returns:
+        res (np.ndarray): Refined coords in shape of batchsize*num_kps*2.
+    """
+    if not isinstance(batch_heatmaps, np.ndarray):
+        batch_heatmaps = batch_heatmaps.cpu().numpy()
+    batch, num_kps, h, w = batch_heatmaps.shape
+    persons = coords.shape[0]
+    if share_heatmap:
+        assert batch == 1
+    else:
+        assert batch == persons
+    for i in range(batch):
+        for j in range(num_kps):
+            mapij = batch_heatmaps[0, j, :, :]
+            mapij = cv2.GaussianBlur(mapij, (3, 3), 0)
+            batch_heatmaps[0, j, :, :] = mapij
+    batch_heatmaps = np.clip(batch_heatmaps, 0.001, 50)
+    batch_heatmaps = np.log(batch_heatmaps)
+    batch_heatmaps = np.transpose(batch_heatmaps, (2, 3, 0, 1)).reshape(
+        (h, w, batch * num_kps))
+    batch_heatmaps_pad = cv2.copyMakeBorder(
+        batch_heatmaps, 1, 1, 1, 1, borderType=cv2.BORDER_REFLECT)
+    batch_heatmaps_pad = np.transpose(
+        batch_heatmaps_pad.reshape((h + 2, w + 2, batch, num_kps)),
+        (2, 3, 0, 1))
+    index = coords[:, :, 0] + 1 + (coords[:, :, 1] + 1) * (w + 2)
+
+    index = index.reshape((persons, -1))
+    if share_heatmap:
+        offset = np.arange(0, num_kps) * (w + 2) * (h + 2)
+        index = index + offset[None, :]
+    else:
+        offset = np.arange(0, batch * num_kps) * (w + 2) * (h + 2)
+        index = index + offset.reshape((persons, -1))
+    index = index.astype(np.int).reshape((-1, 1))
+    batch_heatmaps_pad = batch_heatmaps_pad.reshape((-1, ))
+    i_ = batch_heatmaps_pad[index]
+    ix1 = batch_heatmaps_pad[index + 1]
+    iy1 = batch_heatmaps_pad[index + w + 2]
+    ix1y1 = batch_heatmaps_pad[index + w + 3]
+    ix1_y1_ = batch_heatmaps_pad[index - w - 3]
+    ix1_ = batch_heatmaps_pad[index - 1]
+    iy1_ = batch_heatmaps_pad[index - 2 - w]
+
+    dx = 0.5 * (ix1 - ix1_)
+    dy = 0.5 * (iy1 - iy1_)
+    derivative = np.concatenate([dx, dy], axis=1).reshape(
+        (persons, num_kps, 2, 1))
+    dxx = ix1 - 2 * i_ + ix1_
+    dyy = iy1 - 2 * i_ + iy1_
+    dxy = 0.5 * (ix1y1 - ix1 - iy1 + i_ + i_ - ix1_ - iy1_ + ix1_y1_)
+    hessian = np.concatenate([dxx, dxy, dxy, dyy], axis=1).reshape(
+        (persons, num_kps, 2, 2))
+    for i in range(persons):
+        for j in range(num_kps):
+            hessian_tmp = hessian[i, j, :, :] + \
+                          np.finfo(np.float32).eps * np.eye(2)
+            derivative_tmp = derivative[i, j, :, :]
+            shift = np.matmul(np.linalg.inv(hessian_tmp), derivative_tmp)
```

Comment:
Since it is 2x2, you can directly write the solution of inv, so that the two for loops could be avoided?

Reply:
use this instead~
>hessian = np.linalg.inv(hessian + np.finfo(np.float32).eps * np.eye(2))
 >coords += np.einsum('ijmn,ijnk->ijmk', hessian, derivative).squeeze()

Reply:
This is better, we can do it even better. I mean, eliminate the two outer for i, j loops

Reply:
in this way, the two outer for i,j loops have been removed

Reply:
yeah makes sense.

I'm waiting for a revision before continue :p

Reply:
done